public/app/features/dashboard/dashgrid/DashboardEmpty/DashboardEmptyHooks.ts

```diff
@@ -0,0 +1,84 @@
+import { useMemo } from 'react';
+
+import { locationService } from '@grafana/runtime';
+import { buildPanelEditScene } from 'app/features/dashboard-scene/panel-edit/PanelEditor';
+import { DashboardScene } from 'app/features/dashboard-scene/scene/DashboardScene';
+import { DashboardInteractions } from 'app/features/dashboard-scene/utils/interactions';
+import { useGetResourceRepositoryView } from 'app/features/provisioning/hooks/useGetResourceRepositoryView';
+import { useDispatch, useSelector } from 'app/types/store';
+
+import { setInitialDatasource } from '../../state/reducers';
+import {
+  onCreateNewPanel,
+  onAddLibraryPanel as onAddLibraryPanelImpl,
+  onImportDashboard as onImportDashboardImpl,
+} from '../../utils/dashboard';
+
+import type { Props } from './DashboardEmpty';
+
+export const useOnAddVisualization = ({ dashboard, canCreate }: Props) => {
+  const dispatch = useDispatch();
+  const initialDatasource = useSelector((state) => state.dashboard.initialDatasource);
+  const { isReadOnlyRepo } = useGetResourceRepositoryView({
+    folderName: dashboard instanceof DashboardScene ? dashboard.state.meta.folderUid : dashboard.meta.folderUid,
+  });
```

Comment:
This is duplicated inside all 3 hooks, maybe better placed outside it? (also worth checking that this does not result in multuple duplicate API calls)

Reply:
Pushed up two sets of de-duplication changes to reduce how many times the hooks are being called. Should just be calling this check once now.